codelists/forms.py

```diff
@@ -36,6 +37,38 @@ class Meta:
 SignOffFormSet = forms.modelformset_factory(SignOff, form=SignOffForm, can_delete=True)
 
 
+class CodelistUniquenessMixin:
+    def full_clean(self):
+        """
+        Override full_clean() to validate Name and Slug are unique to a Project.
+
+        The Codelist forms don't include `project` or `slug` fields.  However,
+        Codelists are unique based on the Project, Name, and Slug, and we still
+        want to take advantage of Django's ModelForm validation handling so we
+        can report non-unique errors to the user.
+
+        This method calls ModelForm's full_clean() to copy values from the form
+        to the instance.  Both the Create and Update views pass in an instance
+        (either with just a Project or with the existing Codelist which also
+        has it's Project).  It then generates a Slug before validating
+        uniqueness of the instance and adding errors if appropriate.
+        """
+        # Call super()'s full_clean to populate the Codelist instance with
+        # values from the form.
+        super().full_clean()
+
+        # populate the instance's slug so the validate_unique() check below can
+        # use it to check uniqueness
+        self.instance.slug = slugify(self.instance.name)
+
+        # Validate uniqueness of the instance now it's been populated with
+        # values from the form and the Project instance.
+        try:
+            self.instance.validate_unique()
+        except forms.ValidationError as e:
+            self._update_errors(e)
```

Comment:
Nit: you could call call `self.validate_unique` directly here, which avoids prodding at parts of the private API.

Reply:
`self` here is a ModelForm instance and `validate_unique` looks like it's only exposed as a [Model method](https://docs.djangoproject.com/en/3.1/ref/models/instances/#django.db.models.Model.validate_unique), am I missing something?

Good spot on the private API though, that should be a call to `add_error(None, e)`!

Reply:
It's also a [method on `BaseModelForm`](https://github.com/django/django/blob/master/django/forms/models.py#L413-L422), and does almost exactly what you're doing here.

Reply:
TIL 🎉  Thanks!

Reply:
We can't use `BaseModelForm.validate_unique` because it drops fields not defined on the form (I've added more information in a comment which hopefully gives enough context in 15e56e4), bit frustrating!

I've switch to `add_error(None, e)` to avoid the private API in cb95f37.

Reply:
Thanks.  Don't think it's worth spending any more time on this!